File: src/main/java/com/matag/game/cardinstance/CardListComponent.java
package com.matag.game.cardinstance;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public abstract class CardListComponent {
  protected List<CardInstance> cards = new ArrayList<>();

  public CardInstanceSearch search() {
    return new CardInstanceSearch(cards);
  }

  public List<CardInstance> getCards() {
    return cards;
  }

  public void setCards(List<CardInstance> cards) {
    this.cards = cards;
  }

  public int size() {
    return cards.size();
  }

  public void addCard(CardInstance cardInstance) {
    this.cards.add(cardInstance);
  }

  public void addCards(List<CardInstance> cardInstance) {
    this.cards.addAll(cardInstance);
  }

  public boolean hasCardById(int cardId) {
    return search().withId(cardId).isPresent();
  }

  public CardInstance findCardById(int cardId) {
    return search().withId(cardId).orElseThrow(() -> new RuntimeException("Card with id " + cardId + " not found."));
  }

  public List<CardInstance> extractCardsByIds(List<Integer> cardIds) {
    return cardIds.stream()
      .map(this::findCardById)
      .peek(cardInstance -> cards.remove(cardInstance))
      .collect(Collectors.toList());
  }

  public CardInstance extractCardById(int cardId) {
    return extractCardsByIds(List.of(cardId)).get(0);
  }

  public ArrayList<CardInstance> extractAllCards() {
    var removedCards = new ArrayList<CardInstance>();
    while (!cards.isEmpty()) {
      removedCards.add(cards.remove(0));
    }
    return removedCards;
  }

}

File: src/main/kotlin/com/matag/game/cardinstance/CardListComponent.kt
package com.matag.game.cardinstance

abstract class CardListComponent {
    var cards: List<CardInstance> = listOf()

    fun search(): CardInstanceSearch {
        return CardInstanceSearch(cards)
    }

    fun size(): Int {
        return cards.size
    }

    fun addCard(cardInstance: CardInstance) {
        cards = cards + cardInstance
    }

    fun addCards(cardInstance: List<CardInstance>) {
        cards + cardInstance
    }

    fun hasCardById(cardId: Int): Boolean =
        search().withId(cardId) != null

    fun findCardById(cardId: Int): CardInstance =
        search().withId(cardId) ?: throw RuntimeException("Card with id $cardId not found.")

    fun extractCardsByIds(cardIds: List<Int>): List<CardInstance> =
        cardIds.map { findCardById(it) }
            .also { this.cards -= it }

    fun extractCardById(cardId: Int): CardInstance {
        return extractCardsByIds(listOf(cardId)).single()
    }

    fun extractAllCards(): List<CardInstance> =
        cards.also {
            cards = emptyList()
        }
}
